core-backend/src/env.rs

```diff
@@ -155,8 +163,31 @@ fn store_host_state_mut<Ext>(
         .unwrap_or_else(|| unreachable!("State must be set in `WasmiEnvironment::new`; qed"))
 }
 
+pub type EnvironmentExecutionResult<Ext, PrepareMemoryError> =
+    Result<BackendReport<Ext>, EnvironmentError<PrepareMemoryError>>;
+
 #[derive(Debug, derive_more::Display)]
-pub enum SandboxEnvironmentError {
+pub enum EnvironmentError<PrepareMemoryError: Display> {
+    #[display(fmt = "Actor backend error: {_1}")]
+    Actor(GasAmount, String),
+    #[display(fmt = "System backend error: {_0}")]
+    System(SystemEnvironmentError),
+    #[display(fmt = "Prepare error: {_1}")]
+    PrepareMemory(GasAmount, PrepareMemoryError),
+}
+
+impl<PrepareMemoryError: Display> EnvironmentError<PrepareMemoryError> {
+    pub fn from_infallible(err: EnvironmentError<Infallible>) -> Self {
+        match err {
+            EnvironmentError::System(err) => Self::System(err),
+            EnvironmentError::PrepareMemory(_, err) => match err {},
```

Comment:
```suggestion
            EnvironmentError::PrepareMemory( .. ) => unreachable!("Infallible passed"),
```

Reply:
What's the point to change type-system-proven unreachable to runtime one?